scripts/exact_1d_bayesian_filtering_demo.py

```diff
@@ -0,0 +1,515 @@
+    v = v_rvs(rng_keys[0], shape=(max_iter + 1, num_samples))
+    e = e_rvs(rng_keys[1], shape=(max_iter + 1, num_samples))
+
+    for k in range(1, max_iter + 1):
+        x = inversion_sampling(
+            rng_keys[2],
+            x_grid,
+            p_filter[k - 1, :],
+            num_samples=num_samples,
+        )
+
+        x = f(x, v[k], k - 1)
+
+        # p(xk | y(1:k-1))
+        p_pred = p_pred.at[k, :].set(kde(x_grid, x, kernel_variance))
+        p_pred = p_pred.at[k, :].set(p_pred[k, :] / jnp.sum(p_pred[k, :]))
+
+        # measurement
+        y = h(x, e[k])
+
+        # p(xk | y(1:k))
+        threshold = 3 * jnp.sqrt(kernel_variance)
+        distance = jnp.abs(y[k] - y)
+
+        def update(xi, yi, distance_i):
+            return jnp.where(
+                distance_i < threshold,
+                jsp.stats.norm.pdf(x_grid, xi, kernel_variance)
+                * jsp.stats.norm.pdf(y[k], yi, kernel_variance),
+                0,
+            )
+
+        update_vals = jax.vmap(update)(x, y, distance)
+        p_filter = p_filter.at[k, :].set(p_filter[k, :] + jnp.sum(update_vals, axis=0))
+        p_filter = p_filter.at[k, :].set(p_filter[k, :] / jnp.sum(p_filter[k, :]))
+    return p_filter / delta, p_pred / delta
+
+
+def point_mass_density(
+        y: jnp.DeviceArray, x_grid: jnp.DeviceArray, x0_pdf: Callable,
+        x_pdf: Callable, v_pdf: Callable, e_pdf: Callable,
+        f: Callable, h: Callable):
+    num_grid_points = x_grid.shape[0]
+    max_iter = len(y) - 1
+    delta = x_grid[1] - x_grid[0]
+    X = jnp.tile(x_grid, (num_grid_points, 1))
+
+    p_filter = jnp.zeros(shape=(max_iter + 1, num_grid_points))
+    p_filter = p_filter.at[0, :].set(x0_pdf(x_grid))
+    p_filter = p_filter.at[0, :].set(p_filter[0, :] / sum(p_filter[0, :]))  # normalize
+    p_pred = jnp.ones(shape=(max_iter + 1, num_grid_points)) * jnp.inf
+
+    for k in range(1, max_iter + 1):
+        # p(xk, xk-1 | y(1:k-1))
+        px = x_pdf(k=k - 1, x_new=X.T, x=X, v_pdf=v_pdf, f=f)
+        p_joint = px * p_filter[k - 1, :]
+
+        # p(xk | y(1:k-1))
+        p_pred = p_pred.at[k, :].set(jnp.sum(p_joint, axis=1))
+        p_pred = p_pred.at[k, :].set(p_pred[k, :] / jnp.sum(p_pred[k, :]))
+
+        # p(xk | y(1:k))
+        p_filter = p_filter.at[k, :].set(
+            p_pred[k, :] * y_likelihood(y[k], x_grid, e_pdf, h)
+        )
+        p_filter = p_filter.at[k, :].set(p_filter[k, :] / jnp.sum(p_filter[k, :]))
+
+    p_smooth = jnp.array(p_filter.copy())
+    for k in range(max_iter - 1, -1, -1):
+        # p(xk, xk-1 | y(1:k-1))
+        px = x_pdf(k=k, x_new=X.T, x=X, v_pdf=v_pdf, f=f)
+        px = px * p_smooth[k + 1, :].T / p_pred[k + 1, :].T
+
+        p_smooth = p_smooth.at[k, :].set(jnp.sum(px, axis=1))  # marginalize
+        p_smooth = p_smooth.at[k, :].set(
+            p_smooth[k, :] * p_filter[k, :]
+        )  # multiply p(xk|yk)
+        p_smooth = p_smooth.at[k, :].set(p_smooth[k, :] / jnp.sum(p_smooth[k, :]))
```

Comment:
Try to use [jax.lax.scan](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html)